nwg_panel/icons.py

```diff
@@ -70,10 +70,19 @@ def __process_desktop_file(file_path):
 
 
 def __populate_caches():
-    for d in __get_app_dirs():
+    app_dirs = __get_app_dirs()
+    seen_dirs = set()
+    while app_dirs:
+        d = app_dirs.pop(0)
+        if d in seen_dirs:
+            continue
+        seen_dirs.add(d)
         if os.path.isdir(d):
             for file_name in os.listdir(d):
-                file_path = os.path.join(d, file_name)
+                file_path = os.path.realpath(os.path.join(d, file_name))
+                if os.path.isdir(file_path):
+                    app_dirs.insert(0, file_path)
+                    continue
                 __process_desktop_file(file_path)
```

Comment:
nit: you could add
```
if not os.path.exists(file_path):
    continue
```
to avoid a crash on launch if the directory contains a broken symlink

Reply:
Very good idea. I haven't thought of that case.

Reply:
@pol-rivero I've also been thinking whether or not check if the file has a `.desktop` extension before passing it to `__process_desktop_file` method.

Reply:
I also had the same doubt when writing it originally, by decided not to check it for 2 reasons:
1. It's not necessary. If our code works fine without this restriction, we shouldn't enforce it (see [Robustness principle](https://en.wikipedia.org/wiki/Robustness_principle)).
2. The [Desktop Entry Specification](https://specifications.freedesktop.org/desktop-entry-spec/latest/file-naming.html) technically _does_ allow desktop files to use extensions other than `.desktop` (notice the use of "Desktop entry files **should** have the .desktop extension" instead of "**must**" or equivalent).